### PR TITLE
fix(workspace-changes): record symlink targets without the verbatim prefix

### PR DESCRIPTION
**Pull request by Shxiao101**
## Why

`os.readlink` on Windows reports absolute symlink targets in extended-length form (`\\?\C:\...` for drive paths, `\\?\UNC\server\share` for UNC). The workspace-changes scanner stored that raw spelling in `FileSnapshot.symlink_target`, so workspace-change events carried `\\?\`-prefixed targets that do not compare equal to (or display like) ordinary Windows paths — the same change observed through the events API showed two different spellings of one target depending on how the link was created.

## What changed

- `backend/packages/harness/deerflow/workspace_changes/scanner.py` strips the extended-length prefix when recording a symlink target (both the drive and the `UNC` forms). Relative targets and POSIX readlink output pass through unchanged, so Linux CI behavior is identical.
- The `skills/projection.py` and `skills/review/readers.py` readlink sites are deliberately untouched: projection relies on the raw form for its absolute-symlink rejection check, and neither has a user-facing contract pinned on the spelling.

## Surface area

- [ ] **Frontend UI** — page / component / setting / interaction under `frontend/`
- [x] **Backend API** — endpoint / SSE event / request-response shape under `backend/app` — workspace-change events now report symlink targets in their ordinary spelling (value normalization only; payload shape unchanged)
- [ ] **Agents / LangGraph** — agent node, graph wiring, `langgraph.json`, or prompt change
  - Prompt-layer self-check: for every data source in the new text, what is its trust level, and which channel should it ride? Model-supplied or user-influenceable values belong on the untrusted, sanitized data channel (e.g. the task `HumanMessage`) — never interpolated into framework-owned system text, even neutralized.
- [ ] **Sandbox** — `docker/` or sandboxed execution
- [ ] **Skills** — change under `skills/`
- [ ] **Dependencies** — new/upgraded entry in `backend/pyproject.toml` or `frontend/package.json` (say what it buys us)
- [ ] **Default behavior change** — changes existing behavior without the user opting in (default model, default setting, data shape)
- [ ] **Docs / tests / CI only** — no runtime behavior change

## Bug fix verification

- Test path that reproduces the bug: `backend/tests/test_workspace_changes.py::test_scan_workspace_roots_captures_symlinks_as_metadata_only_stubs` (and the sibling `test_compare_snapshots_classifies_symlink_replacing_file_as_symlink_created`) — pre-fix assertion failure on a Windows host with symlink privilege: `assert '\\?\C:\U...de-target.txt' == 'C:\Users\S...de-target.txt'`.
- Did it go red on `main` and green on this branch? **Yes** on a Windows 11 host with symlink privilege: both failed on `main`, 28/28 pass on this branch. Note these tests `skip` where symlink creation is not permitted (stock Windows without Developer Mode, including default GitHub `windows-latest` runners); on Linux CI the tests run and the normalization is a no-op. The helper was additionally unit-checked against drive, UNC, relative, POSIX, and 300-character targets.

## Validation

Windows 11 host, Git Bash (`make` is unavailable here, as also reported in #5177):

- `uv run pytest tests/test_workspace_changes.py` — 28 passed
- `uv run ruff check .` and `uv run ruff format --check .` — clean (backend-wide)

## AI assistance

**Tool(s) used:** ZCode (GLM agent)

**How you used it:** triaged the Windows-host test-suite failures, drafted the normalization helper and PR text; I reviewed the full diff line by line before submitting.

- [x] I've read and understand every line of this change and take responsibility for it — it's not unreviewed AI output.

### CODE REVIEW

File: backend/packages/harness/deerflow/workspace_changes/scanner.py [hunk 1]
```diff
@@ -259,6 +259,21 @@ def _snapshot_file(
     )
 
 
+def _normalize_symlink_target(target: str) -> str:
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/workspace_changes/scanner.py:262]: The two tests cited in the PR description don't cover this branch in CI: on Linux `os.readlink` returns the target verbatim, so `test_scan_workspace_roots_captures_symlinks_as_metadata_only_stubs` and `test_compare_snapshots_classifies_symlink_replacing_file_as_symlink_created` pass both before and after this change (and they `skip` on stock `windows-latest` runners without symlink privilege). So this PR ships untested code, even though the PR body mentions the helper was "unit-checked" ad-hoc against drive/UNC/relative/POSIX/300-char targets.

Since `_normalize_symlink_target` is pure and platform-independent, please commit that unit check to `backend/tests/test_workspace_changes.py` — e.g. drive, `\\?\UNC\` -> `\\`, relative, plain POSIX, and a target that merely *contains* `\\?\` mid-string (must be left alone). That gives every platform real coverage of the new code instead of relying on a Windows host with symlink privilege.

File: backend/packages/harness/deerflow/workspace_changes/scanner.py [hunk 2]
```diff
@@ -259,6 +259,21 @@ def _snapshot_file(
...
+    """Strip the Windows extended-length prefix from a symlink target.
+
+    ``os.readlink`` on Windows reports absolute targets in extended-length
+    form (``\\\\?\\C:\\...`` or ``\\\\?\\UNC\\server\\share``). Recorded targets
+    are surfaced in workspace-change events and compared against ordinary
+    paths, so keep the plain spelling; on POSIX this is a no-op.
+    """
+    if target.startswith("\\\\?\\UNC\\"):
+        return "\\\\" + target[len("\\\\?\\UNC\\") :]
+    if target.startswith("\\\\?\\"):
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/workspace_changes/scanner.py]: Nit: as written the prefix strip also applies on POSIX, so the docstring's "on POSIX this is a no-op" is only true for typical `readlink` output. `readlink(2)` returns the literal string the link was created with, and backslash is a valid filename byte on Linux — a workspace containing a symlink whose target string literally starts with `\\?\` (achievable through sandbox-created links) would get that target rewritten to `\\` + remainder instead of being recorded verbatim.

Given this module's otherwise-defensive posture toward symlinks, consider gating the whole helper on `os.name == "nt"` and returning `target` unchanged otherwise. It makes the docstring claim literally correct and the POSIX path provably identity.

**Review by willem-bd — COMMENTED** [review]
Reviewed at 31d201e5a807d9c2d01da5b93c0625d1176e2593. The normalization itself is correct: both the drive form (`\\?\C:\...` -> `C:\...`) and the UNC form (`\\?\UNC\srv\share` -> `\\srv\share`) round-trip properly, relative/POSIX targets pass through, and the `else:` branch keeps the `OSError` -> `None` path intact. I also confirmed it cannot cause spurious change reports, since `_same_file` in `diff.py` keys on sha256/size/mtime and never compares `symlink_target`. Leaving `projection.py` untouched is sound (`Path(target).is_absolute()` is True for every one of the four spellings above). Two follow-ups: the new branch has no committed test coverage, and the helper would ideally be gated to Windows so the docstring's "POSIX no-op" claim is literally true.

**Comment by Shxiao101** [comment]
Both points addressed in one commit:

1. The helper is now gated on `os.name == "nt"`, so the POSIX path is a provable identity — a target string that merely starts with `\\?\` on a Linux host (legal filename bytes) is recorded verbatim, and the docstring states exactly that.
2. The unit checks are now committed in `backend/tests/test_workspace_changes.py` (5 cases): drive-prefix strip, `\\?\UNC\` → `\\`, relative and plain-POSIX verbatim, a mid-string `\\?\` left untouched on both spellings, and an off-Windows identity case — so the new branch has real coverage on every platform, not just a symlink-privileged Windows host.

(Edit: fixed backslash spellings in this reply.)

File: backend/tests/test_workspace_changes.py [hunk 3]
```diff
@@ -741,3 +746,31 @@ class FakeRequest:
     assert response["available"] is True
     assert response["files"] == []
     assert calls["event_types"] == ["workspace_changes"]
+
+
+def test_normalize_symlink_target_strips_extended_length_drive_prefix():
+    assert _normalize_symlink_target(r"\\?\C:\Users\u1\target.txt") == r"C:\Users\u1\target.txt"
```
Comment (willem-bd) [inline backend/tests/test_workspace_changes.py:755]: Adding the `os.name != "nt"` gate (the previous round's suggestion) makes these two new tests fail on every POSIX host — and `backend-unit-tests.yml` runs only on `ubuntu-latest`, so CI is red:

```
FAILED tests/test_workspace_changes.py::test_normalize_symlink_target_strips_extended_length_drive_prefix
FAILED tests/test_workspace_changes.py::test_normalize_symlink_target_strips_extended_length_unc_prefix
2 failed, 3 passed in 0.52s
```

With the gate, `_normalize_symlink_target` returns its input verbatim off-Windows, so neither assertion can hold there. Mirror `test_normalize_symlink_target_is_identity_off_windows` and force the Windows path in these two:

```python
def test_normalize_symlink_target_strips_extended_length_drive_prefix(monkeypatch):
    monkeypatch.setattr(os, "name", "nt")
    assert _normalize_symlink_target(r"\\?\C:\Users\u1\target.txt") == r"C:\Users\u1\target.txt"
```

Same for `test_normalize_symlink_target_leaves_relative_and_plain_posix_targets_verbatim` and `test_normalize_symlink_target_leaves_mid_string_prefix_verbatim`: on POSIX they pass vacuously, because the helper returns before reaching the prefix checks. Patching `os.name` to `"nt"` there too is what keeps the strip branch (and its fallthrough) covered on every platform, which was the point of the earlier coverage request.

**Review by willem-bd — COMMENTED** [review]
Reviewed at dd873e229. Both findings from the 31d201e5 round are addressed in code: the helper is now gated on `os.name == "nt"` (POSIX path is a provable identity), and the pure-function unit tests are committed. One new blocking issue from combining the two — the gate makes two of those new tests fail on POSIX, the only OS the backend CI suite runs on. Details inline.

**Comment by Shxiao101** [comment]
Good catch — the gate made those two assertions impossible on the ubuntu CI, and I only ran the suite on Windows. Both prefix-strip tests now force `os.name = "nt"` via monkeypatch (mirroring the off-Windows identity test), so each of the five cases pins exactly one platform's contract: 03ef585e.

File: backend/packages/harness/deerflow/workspace_changes/scanner.py [hunk 4]
```diff
@@ -259,6 +259,28 @@ def _snapshot_file(
...
+    hosts. ``readlink(2)`` returns the literal string the link was created
+    with, and backslash is a valid filename byte on Linux — a target string
+    that merely starts with ``\\\\?\\`` there must be recorded verbatim.
+    """
+    if os.name != "nt":
+        return target
+    if target.startswith("\\\\?\\UNC\\"):
+        return "\\\\" + target[len("\\\\?\\UNC\\") :]
+    if target.startswith("\\\\?\\"):
+        return target[len("\\\\?\\") :]
```
Comment (willem-bd) [inline backend/packages/harness/deerflow/workspace_changes/scanner.py]: [P2] Preserve prefixes for non-drive Windows paths

This catch-all branch also strips the prefix from volume-GUID targets such as `\\?\Volume{01234567-89ab-cdef-0123-456789abcdef}\folder\target.txt`, producing `Volume{01234567-89ab-cdef-0123-456789abcdef}\folder\target.txt`. That is a relative-looking path rather than an equivalent spelling of the absolute target, so workspace-change metadata loses the target's namespace. Volume-GUID paths are a valid Windows form: https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-volume.

Please restrict this branch to extended drive-letter paths, preserve other namespace forms, and add a volume-GUID regression test. A Windows-mode helper probe confirms that `ntpath.isabs()` changes from true to false for the example above; native Windows symlink creation was not tested.

**Review by willem-bd — COMMENTED** [review]
Reviewed at 03ef585e. All 33 workspace-change tests passed. One Windows namespace edge case remains, verified with a Windows-mode helper probe; native Windows symlink creation was not tested.

**Comment by Shxiao101** [comment]
Fixed in a7d138c5: the catch-all branch now strips only extended drive-letter paths — it requires a drive letter, a colon, and a path separator after the `\\?\` prefix — so volume-GUID targets (and any other namespace form, plus degenerate spellings like `\\?\C:`) are recorded verbatim. Added the volume-GUID regression you asked for alongside degenerate-prefix cases, all Windows-mode like the rest.

**Review by WillemJiang — APPROVED** [review]
